File: Linked List/61.Rotate List/Solution.py
class Solution:
    # @param head, a ListNode
    # @param k, an integer
    # @return a ListNode
    def rotateRight(self, head, k):
        if not head or not head.next:
            return head
        if k == 0:
            return head
        p1 = head
        while k > 0:
            p1 = p1.next
            if not p1: p1 = head
            k -= 1
        p2 = head
        while p1.next != None:
            p1 = p1.next
            p2 = p2.next
        p1.next = head
        head = p2.next
        p2.next = None
        return head

File: Linked List/61.Rotate List/Solution.java
/**
 * 对链表进行右移k步，k为非负数
 * 题目地址：https://leetcode.com/problems/rotate-list/
**/
public class Solution {
    public ListNode rotateRight(ListNode head, int k) {
        if(head == null || head.next == null || k <= 0) return head;
        ListNode pHead = new ListNode(0);
        pHead.next = head;
        int len = 0;
        ListNode p = pHead;
        while(p.next != null) {  //找到链表长度，方便从头找待移结点。
            p = p.next;
            len++;
        }
        ListNode q = pHead;
        for(int i = 0; i < len - k % len; i++) {  //防止k过大导致TLE，取余。
            q = q.next;
        }
        p.next = pHead.next;
        pHead.next = q.next;
        q.next = null;
        return pHead.next;
    }
}